src/back/msl/writer.rs

```diff
@@ -1721,9 +1721,56 @@ impl<W: Write> Writer<W> {
                     self.put_expression(arg, context, true)?;
                     write!(self.out, ") + 1) % 33) - 1)")?;
                 } else if fun == Mf::FindMsb {
-                    write!(self.out, "((({NAMESPACE}::clz(")?;
-                    self.put_expression(arg, context, true)?;
-                    write!(self.out, ") + 1) % 33) - 1)")?
+                    let inner = context.resolve_type(arg);
+
+                    if let Some(crate::ScalarKind::Uint) = inner.scalar_kind() {
+                        write!(self.out, "((({NAMESPACE}::clz(")?;
+                        self.put_expression(arg, context, true)?;
+                        write!(self.out, ") + 1) % 33) - 1)")?;
+                    }
+
+                    match *inner {
+                        crate::TypeInner::Vector {
+                            size,
+                            kind: crate::ScalarKind::Sint,
+                            ..
+                        } => {
+                            let s = back::vector_size_str(size);
+                            write!(self.out, "{NAMESPACE}::select(int{s}(")?;
+                            write!(self.out, "{NAMESPACE}::floor(")?;
+                            write!(self.out, "{NAMESPACE}::log2(float{s}(")?;
+                            write!(self.out, "{NAMESPACE}::select(")?;
+                            self.put_expression(arg, context, false)?;
+                            write!(self.out, ", ~")?;
+                            self.put_expression(arg, context, false)?;
+                            write!(self.out, ", ")?;
+                            self.put_expression(arg, context, false)?;
+                            write!(self.out, " < 0))))), int{s}(-1), ")?;
+                            self.put_expression(arg, context, false)?;
+                            write!(self.out, " == 0 || ")?;
+                            self.put_expression(arg, context, false)?;
```

Comment:
One problem with using `arg` repeatedly here is that it will get repeated in the source code. If it happens to be long, that could be awkward. (If it contains further `firstLeadingBit` calls, the expansion could be exponential - unlikely, but still unpleasant.)

Instead, if we do need to use `arg` repeatedly like this, we should recognize `findMsb` in `Writer::update_expressions_to_bake`, as we do `crate::MathFunction::Dot`.

Reply:
This is handled further along in the file